File: backend/src/AgentTeam/Controllers/ScheduledWorkflowController.php
<?php

declare(strict_types=1);

namespace AgentTeam\Controllers;

use AgentTeam\Services\ScheduledWorkflowService;
use AgentTeam\Services\WorkflowRepository;
use PDO;

/**
 * Scheduled Workflow Controller
 *
 * Handles REST API endpoints for managing scheduled workflow executions.
 */
class ScheduledWorkflowController
{
    private PDO $db;
    private array $config;
    private ScheduledWorkflowService $scheduleService;
    private WorkflowRepository $workflowRepository;

    public function __construct(PDO $db, array $config)
    {
        $this->db = $db;
        $this->config = $config;
        $this->scheduleService = new ScheduledWorkflowService($db);
        $this->workflowRepository = new WorkflowRepository($db);
    }

    /**
     * GET /api/v1/schedules
     * List all scheduled workflows for the authenticated user
     */
    public function index(array $request): array
    {
        $userId = $request['user_id'] ?? 0;

        if (!$userId) {
            return [
                'success' => false,
                'error' => 'Authentication required',
                'status_code' => 401
            ];
        }

        try {
            $schedules = $this->scheduleService->findByUser($userId);

            return [
                'success' => true,
                'data' => $schedules,
                'count' => count($schedules),
                'status_code' => 200
            ];
        } catch (\Exception $e) {
            return [
                'success' => false,
                'error' => $e->getMessage(),
                'status_code' => 500
            ];
        }
    }

    /**
     * POST /api/v1/schedules
     * Create a new scheduled workflow
     */
    public function create(array $request): array
    {
        $userId = $request['user_id'] ?? 0;
        $body = $request['body'] ?? [];

        if (!$userId) {
            return [
                'success' => false,
                'error' => 'Authentication required',
                'status_code' => 401
            ];
        }

        // Validate required fields
        if (empty($body['workflow_id'])) {
            return [
                'success' => false,
                'error' => 'Workflow ID is required',
                'status_code' => 400
            ];
        }

        if (empty($body['scheduled_time'])) {
            return [
                'success' => false,
                'error' => 'Scheduled time is required',
                'status_code' => 400
            ];
        }

        // Verify the workflow exists and belongs to the user
        $workflow = $this->workflowRepository->findById((int) $body['workflow_id']);
        if (!$workflow || $workflow->getUserId() !== $userId) {
            return [
                'success' => false,
                'error' => 'Workflow not found',
                'status_code' => 404
            ];
        }

        // Validate repeat_type if provided
        $validRepeatTypes = ['none', 'hourly', 'daily', 'weekly', 'monthly'];
        if (!empty($body['repeat_type']) && !in_array($body['repeat_type'], $validRepeatTypes)) {
            return [
                'success' => false,
                'error' => 'Invalid repeat type. Must be one of: ' . implode(', ', $validRepeatTypes),
                'status_code' => 400
            ];
        }

        try {
            $schedule = $this->scheduleService->create([
                'user_id' => $userId,
                'workflow_id' => (int) $body['workflow_id'],
                'input_prompt' => $body['input_prompt'] ?? null,
                'scheduled_time' => $body['scheduled_time'],
                'repeat_type' => $body['repeat_type'] ?? 'none',
                'repeat_interval' => (int) ($body['repeat_interval'] ?? 1)
            ]);

            return [
                'success' => true,
                'data' => $schedule,
                'message' => 'Schedule created successfully',
                'status_code' => 201
            ];
        } catch (\Exception $e) {
            return [
                'success' => false,
                'error' => $e->getMessage(),
                'status_code' => 500
            ];
        }
    }

    /**
     * GET /api/v1/schedules/{id}
     * Get a specific scheduled workflow
     */
    public function show(array $request): array
    {
        $userId = $request['user_id'] ?? 0;
        $id = (int) ($request['params']['id'] ?? 0);

        if (!$userId) {
            return [
                'success' => false,
                'error' => 'Authentication required',
                'status_code' => 401
            ];
        }

        try {
            $schedule = $this->scheduleService->findById($id);

            if (!$schedule || $schedule['user_id'] !== $userId) {
                return [
                    'success' => false,
                    'error' => 'Schedule not found',
                    'status_code' => 404
                ];
            }

            return [
                'success' => true,
                'data' => $schedule,
                'status_code' => 200
            ];
        } catch (\Exception $e) {
            return [
                'success' => false,
                'error' => $e->getMessage(),
                'status_code' => 500
            ];
        }
    }

    /**
     * PUT /api/v1/schedules/{id}
     * Update a scheduled workflow
     */
    public function update(array $request): array
    {
        $userId = $request['user_id'] ?? 0;
        $id = (int) ($request['params']['id'] ?? 0);
        $body = $request['body'] ?? [];

        if (!$userId) {
            return [
                'success' => false,
                'error' => 'Authentication required',
                'status_code' => 401
            ];
        }

        // Validate repeat_type if provided
        if (!empty($body['repeat_type'])) {
            $validRepeatTypes = ['none', 'hourly', 'daily', 'weekly', 'monthly'];
            if (!in_array($body['repeat_type'], $validRepeatTypes)) {
                return [
                    'success' => false,
                    'error' => 'Invalid repeat type',
                    'status_code' => 400
                ];
            }
        }

        try {
            $schedule = $this->scheduleService->update($id, $userId, $body);

            if (!$schedule) {
                return [
                    'success' => false,
                    'error' => 'Schedule not found',
                    'status_code' => 404
                ];
            }

            return [
                'success' => true,
                'data' => $schedule,
                'message' => 'Schedule updated successfully',
                'status_code' => 200
            ];
        } catch (\Exception $e) {
            return [
                'success' => false,
                'error' => $e->getMessage(),
                'status_code' => 500
            ];
        }
    }

    /**
     * DELETE /api/v1/schedules/{id}
     * Delete a scheduled workflow
     */
    public function destroy(array $request): array
    {
        $userId = $request['user_id'] ?? 0;
        $id = (int) ($request['params']['id'] ?? 0);

        if (!$userId) {
            return [
                'success' => false,
                'error' => 'Authentication required',
                'status_code' => 401
            ];
        }

        try {
            $deleted = $this->scheduleService->delete($id, $userId);

            if (!$deleted) {
                return [
                    'success' => false,
                    'error' => 'Schedule not found',
                    'status_code' => 404
                ];
            }

            return [
                'success' => true,
                'message' => 'Schedule deleted successfully',
                'status_code' => 200
            ];
        } catch (\Exception $e) {
            return [
                'success' => false,
                'error' => $e->getMessage(),
                'status_code' => 500
            ];
        }
    }

    /**
     * POST /api/v1/schedules/{id}/pause
     * Pause a scheduled workflow
     */
    public function pause(array $request): array
    {
        $userId = $request['user_id'] ?? 0;
        $id = (int) ($request['params']['id'] ?? 0);

        if (!$userId) {
            return [
                'success' => false,
                'error' => 'Authentication required',
                'status_code' => 401
            ];
        }

        try {
            $paused = $this->scheduleService->pause($id, $userId);

            if (!$paused) {
                return [
                    'success' => false,
                    'error' => 'Schedule not found or cannot be paused',
                    'status_code' => 404
                ];
            }

            return [
                'success' => true,
                'message' => 'Schedule paused successfully',
                'status_code' => 200
            ];
        } catch (\Exception $e) {
            return [
                'success' => false,
                'error' => $e->getMessage(),
                'status_code' => 500
            ];
        }
    }

    /**
     * POST /api/v1/schedules/{id}/resume
     * Resume a paused scheduled workflow
     */
    public function resume(array $request): array
    {
        $userId = $request['user_id'] ?? 0;
        $id = (int) ($request['params']['id'] ?? 0);

        if (!$userId) {
            return [
                'success' => false,
                'error' => 'Authentication required',
                'status_code' => 401
            ];
        }

        try {
            $resumed = $this->scheduleService->resume($id, $userId);

            if (!$resumed) {
                return [
                    'success' => false,
                    'error' => 'Schedule not found or not paused',
                    'status_code' => 404
                ];
            }

            return [
                'success' => true,
                'message' => 'Schedule resumed successfully',
                'status_code' => 200
            ];
        } catch (\Exception $e) {
            return [
                'success' => false,
                'error' => $e->getMessage(),
                'status_code' => 500
            ];
        }
    }

    /**
     * GET /api/v1/schedules/stats
     * Get scheduling statistics for the user
     */
    public function stats(array $request): array
    {
        $userId = $request['user_id'] ?? 0;

        if (!$userId) {
            return [
                'success' => false,
                'error' => 'Authentication required',
                'status_code' => 401
            ];
        }

        try {
            $stats = $this->scheduleService->getStats($userId);

            return [
                'success' => true,
                'data' => $stats,
                'status_code' => 200
            ];
        } catch (\Exception $e) {
            return [
                'success' => false,
                'error' => $e->getMessage(),
                'status_code' => 500
            ];
        }
    }

    /**
     * GET /api/v1/workflows/{id}/schedules
     * Get all schedules for a specific workflow
     */
    public function byWorkflow(array $request): array
    {
        $userId = $request['user_id'] ?? 0;
        $workflowId = (int) ($request['params']['id'] ?? 0);

        if (!$userId) {
            return [
                'success' => false,
                'error' => 'Authentication required',
                'status_code' => 401
            ];
        }

        try {
            $schedules = $this->scheduleService->findByWorkflow($workflowId, $userId);

            return [
                'success' => true,
                'data' => $schedules,
                'count' => count($schedules),
                'status_code' => 200
            ];
        } catch (\Exception $e) {
            return [
                'success' => false,
                'error' => $e->getMessage(),
                'status_code' => 500
            ];
        }
    }
}

File: backend_python/app/agent_team/controllers/scheduled_workflow_controller.py
"""Port of backend/src/AgentTeam/Controllers/ScheduledWorkflowController.php
(436 lines).

CRUD + lifecycle (pause/resume) for scheduled workflow runs, plus
per-workflow listing and aggregate stats. Every method delegates to
`ScheduledWorkflowService` (Task 1, scheduled_workflow_service.py) for the
actual DB work; this controller only does auth/validation/ownership and
response shaping, exactly like PHP. Same method names, same validation
strings, same status codes.

Ported: index, create, show, update, destroy, pause, resume, stats,
byWorkflow. No private helpers in the PHP source for this class.
"""
from __future__ import annotations

from app.agent_team.services.scheduled_workflow_service import ScheduledWorkflowService
from app.agent_team.services.workflow_repository import WorkflowRepository
from app.support.phpcompat import php_empty, php_intval

_VALID_REPEAT_TYPES = ('none', 'hourly', 'daily', 'weekly', 'monthly')


class ScheduledWorkflowController:
    def __init__(self, db, config):
        """PHP 23-29."""
        self.db = db
        self.config = config
        self.scheduleService = ScheduledWorkflowService(db)
        self.workflowRepository = WorkflowRepository(db)

    # ========================================================================
    # GET /api/v1/schedules
    # ========================================================================

    def index(self, request) -> dict:
        """PHP 35-63."""
        userId = request['user_id'] if request.get('user_id') is not None else 0

        if not userId:
            return {'success': False, 'error': 'Authentication required', 'status_code': 401}

        try:
            schedules = self.scheduleService.findByUser(userId)
            return {'success': True, 'data': schedules, 'count': len(schedules), 'status_code': 200}
        except Exception as e:  # noqa: BLE001 -- mirrors PHP `catch (\Exception $e)`
            return {'success': False, 'error': str(e), 'status_code': 500}

    # ========================================================================
    # POST /api/v1/schedules
    # ========================================================================

    def create(self, request) -> dict:
        """PHP 69-142."""
        userId = request['user_id'] if request.get('user_id') is not None else 0
        body = request['body'] if request.get('body') is not None else {}

        if not userId:
            return {'success': False, 'error': 'Authentication required', 'status_code': 401}

        if php_empty(body.get('workflow_id')):
            return {'success': False, 'error': 'Workflow ID is required', 'status_code': 400}

        if php_empty(body.get('scheduled_time')):
            return {'success': False, 'error': 'Scheduled time is required', 'status_code': 400}

        # Verify the workflow exists and belongs to the user
        workflow = self.workflowRepository.findById(php_intval(body['workflow_id']))
        if not workflow or workflow.getUserId() != userId:
            return {'success': False, 'error': 'Workflow not found', 'status_code': 404}

        # Validate repeat_type if provided
        repeatType = body.get('repeat_type')
        if not php_empty(repeatType) and repeatType not in _VALID_REPEAT_TYPES:
            return {
                'success': False,
                'error': 'Invalid repeat type. Must be one of: ' + ', '.join(_VALID_REPEAT_TYPES),
                'status_code': 400,
            }

        try:
            schedule = self.scheduleService.create({
                'user_id': userId,
                'workflow_id': php_intval(body['workflow_id']),
                'input_prompt': body.get('input_prompt'),
                'scheduled_time': body['scheduled_time'],
                'repeat_type': body['repeat_type'] if body.get('repeat_type') is not None else 'none',
                'repeat_interval': php_intval(body['repeat_interval']) if body.get('repeat_interval') is not None else 1,
            })

            return {
                'success': True,
                'data': schedule,
                'message': 'Schedule created successfully',
                'status_code': 201,
            }
        except Exception as e:  # noqa: BLE001
            return {'success': False, 'error': str(e), 'status_code': 500}

    # ========================================================================
    # GET /api/v1/schedules/{id}
    # ========================================================================

    def show(self, request, id: int = 0) -> dict:
        """PHP 148-184."""
        userId = request['user_id'] if request.get('user_id') is not None else 0
        scheduleId = php_intval(id)

        if not userId:
            return {'success': False, 'error': 'Authentication required', 'status_code': 401}

        try:
            schedule = self.scheduleService.findById(scheduleId)

            if not schedule or schedule['user_id'] != userId:
                return {'success': False, 'error': 'Schedule not found', 'status_code': 404}

            return {'success': True, 'data': schedule, 'status_code': 200}
        except Exception as e:  # noqa: BLE001
            return {'success': False, 'error': str(e), 'status_code': 500}

    # ========================================================================
    # PUT /api/v1/schedules/{id}
    # ========================================================================

    def update(self, request, id: int = 0) -> dict:
        """PHP 190-240."""
        userId = request['user_id'] if request.get('user_id') is not None else 0
        scheduleId = php_intval(id)
        body = request['body'] if request.get('body') is not None else {}

        if not userId:
            return {'success': False, 'error': 'Authentication required', 'status_code': 401}

        # Validate repeat_type if provided
        if not php_empty(body.get('repeat_type')):
            if body['repeat_type'] not in _VALID_REPEAT_TYPES:
                return {'success': False, 'error': 'Invalid repeat type', 'status_code': 400}

        try:
            schedule = self.scheduleService.update(scheduleId, userId, body)

            if not schedule:
                return {'success': False, 'error': 'Schedule not found', 'status_code': 404}

            return {
                'success': True,
                'data': schedule,
                'message': 'Schedule updated successfully',
                'status_code': 200,
            }
        except Exception as e:  # noqa: BLE001
            return {'success': False, 'error': str(e), 'status_code': 500}

    # ========================================================================
    # DELETE /api/v1/schedules/{id}
    # ========================================================================

    def destroy(self, request, id: int = 0) -> dict:
        """PHP 246-282."""
        userId = request['user_id'] if request.get('user_id') is not None else 0
        scheduleId = php_intval(id)

        if not userId:
            return {'success': False, 'error': 'Authentication required', 'status_code': 401}

        try:
            deleted = self.scheduleService.delete(scheduleId, userId)

            if not deleted:
                return {'success': False, 'error': 'Schedule not found', 'status_code': 404}

            return {'success': True, 'message': 'Schedule deleted successfully', 'status_code': 200}
        except Exception as e:  # noqa: BLE001
            return {'success': False, 'error': str(e), 'status_code': 500}

    # ========================================================================
    # POST /api/v1/schedules/{id}/pause
    # ========================================================================

    def pause(self, request, id: int = 0) -> dict:
        """PHP 288-324."""
        userId = request['user_id'] if request.get('user_id') is not None else 0
        scheduleId = php_intval(id)

        if not userId:
            return {'success': False, 'error': 'Authentication required', 'status_code': 401}

        try:
            paused = self.scheduleService.pause(scheduleId, userId)

            if not paused:
                return {
                    'success': False,
                    'error': 'Schedule not found or cannot be paused',
                    'status_code': 404,
                }

            return {'success': True, 'message': 'Schedule paused successfully', 'status_code': 200}
        except Exception as e:  # noqa: BLE001
            return {'success': False, 'error': str(e), 'status_code': 500}

    # ========================================================================
    # POST /api/v1/schedules/{id}/resume
    # ========================================================================

    def resume(self, request, id: int = 0) -> dict:
        """PHP 330-366."""
        userId = request['user_id'] if request.get('user_id') is not None else 0
        scheduleId = php_intval(id)

        if not userId:
            return {'success': False, 'error': 'Authentication required', 'status_code': 401}

        try:
            resumed = self.scheduleService.resume(scheduleId, userId)

            if not resumed:
                return {
                    'success': False,
                    'error': 'Schedule not found or not paused',
                    'status_code': 404,
                }

            return {'success': True, 'message': 'Schedule resumed successfully', 'status_code': 200}
        except Exception as e:  # noqa: BLE001
            return {'success': False, 'error': str(e), 'status_code': 500}

    # ========================================================================
    # GET /api/v1/schedules/stats
    # ========================================================================

    def stats(self, request) -> dict:
        """PHP 372-399."""
        userId = request['user_id'] if request.get('user_id') is not None else 0

        if not userId:
            return {'success': False, 'error': 'Authentication required', 'status_code': 401}

        try:
            stats = self.scheduleService.getStats(userId)
            return {'success': True, 'data': stats, 'status_code': 200}
        except Exception as e:  # noqa: BLE001
            return {'success': False, 'error': str(e), 'status_code': 500}

    # ========================================================================
    # GET /api/v1/workflows/{id}/schedules
    # ========================================================================

    def byWorkflow(self, request, id: int = 0) -> dict:
        """PHP 405-434."""
        userId = request['user_id'] if request.get('user_id') is not None else 0
        workflowId = php_intval(id)

        if not userId:
            return {'success': False, 'error': 'Authentication required', 'status_code': 401}

        try:
            schedules = self.scheduleService.findByWorkflow(workflowId, userId)
            return {'success': True, 'data': schedules, 'count': len(schedules), 'status_code': 200}
        except Exception as e:  # noqa: BLE001
            return {'success': False, 'error': str(e), 'status_code': 500}
